plugins/network-elements/nuage-vsp/src/com/cloud/network/element/NuageVspElement.java

```diff
@@ -205,6 +250,75 @@ public boolean implement(Network network, NetworkOffering offering, DeployDestin
             return false;
         }
 
+        final boolean egressDefaultPolicy = offering.getEgressDefaultPolicy();
+        Domain networkDomain = _domainDao.findById(network.getDomainId());
+        boolean isFirewallServiceSupported = _networkModel.areServicesSupportedByNetworkOffering(offering.getId(), Service.Firewall);
+        List<String> dnsServers = _nuageVspManager.getDnsDetails(network);
+
+        boolean isL2Network = false, isL3Network = false, isShared = false;
+        String subnetUuid = network.getUuid();
+        if (offering.getGuestType() == Network.GuestType.Shared) {
+            isShared = true;
+            subnetUuid = networkDomain.getUuid();
+        } else if (_ntwkOfferingSrvcDao.areServicesSupportedByNetworkOffering(offering.getId(), Service.SourceNat)
+                || _ntwkOfferingSrvcDao.areServicesSupportedByNetworkOffering(offering.getId(), Service.StaticNat)
+                || _ntwkOfferingSrvcDao.areServicesSupportedByNetworkOffering(offering.getId(), Service.Connectivity)) {
+            isL3Network = true;
+        } else {
+            isL2Network = true;
+        }
+
+        String preConfiguredDomainTemplateName = getPreConfiguredDomainTemplateName(network, offering);
+        List<FirewallRuleVO> firewallIngressRulesToApply =
+                _firewallRulesDao.listByNetworkPurposeTrafficType(network.getId(), FirewallRule.Purpose.Firewall, FirewallRule.TrafficType.Ingress);
+        for (FirewallRuleVO rule : firewallIngressRulesToApply) {
+            // load cidrs if any
+            rule.setSourceCidrList(_firewallRulesCidrsDao.getSourceCidrs(rule.getId()));
+        }
+
+        List<FirewallRuleVO> firewallEgressRulesToApply =
+                _firewallRulesDao.listByNetworkPurposeTrafficType(network.getId(), FirewallRule.Purpose.Firewall, FirewallRule.TrafficType.Egress);
+        for (FirewallRuleVO rule : firewallEgressRulesToApply) {
+            // load cidrs if any
+            rule.setSourceCidrList(_firewallRulesCidrsDao.getSourceCidrs(rule.getId()));
+        }
+
+        List<Map<String, Object>> ingressFirewallRules = Lists.transform(firewallIngressRulesToApply, new Function<FirewallRuleVO, Map<String, Object>>() {
+            @Override
+            public Map<String, Object> apply(FirewallRuleVO firewallRuleVO) {
+                return getACLRuleDetails(firewallRuleVO, egressDefaultPolicy);
+            }
+        });
+
+        List<Map<String, Object>> egressFirewallRules = Lists.transform(firewallEgressRulesToApply, new Function<FirewallRuleVO, Map<String, Object>>() {
```

Comment:
This ``Function`` is a duplicate of the one declared on line #286.  Extract to ``private static final`` inner class to remove duplication and make the code more readable.